src/commonMain/kotlin/fr/acinq/lightning/blockchain/electrum/ElectrumMiniWallet.kt

```diff
@@ -0,0 +1,162 @@
+    }
+
+    private val outPoint2Address = addresses.entries.flatMap { entry -> entry.value.map { it.outPoint to entry.key } }.toMap()
+
+    /** Sign the inputs owned by this wallet (only works for P2WPKH scripts) */
+    fun sign(tx: Transaction): Transaction {
+        return tx.txIn.foldIndexed(tx) { index, wipTx, txIn ->
+            val witness = outPoint2Address[txIn.outPoint]?.let { address ->
+                addresses[address]?.find { it.outPoint == txIn.outPoint }?.let { utxo ->
+                    privateKeys[address]?.let { privateKey ->
+                        // mind this: the pubkey script used for signing is not the prevout pubscript (which is just a push
+                        // of the pubkey hash), but the actual script that is evaluated by the script engine, in this case a PAY2PKH script
+                        val publicKey = privateKey.publicKey()
+                        val pubKeyScript = Script.pay2pkh(publicKey)
+                        val sig = Transaction.signInput(
+                            tx,
+                            index,
+                            pubKeyScript,
+                            SigHash.SIGHASH_ALL,
+                            utxo.value.sat,
+                            SigVersion.SIGVERSION_WITNESS_V0,
+                            privateKey
+                        )
+                        ScriptWitness(listOf(sig.byteVector(), publicKey.value))
+                    }
+                }
+            }
+            when (witness) {
+                is ScriptWitness ->
+                    // update the signature for the corresponding input
+                    wipTx.updateWitness(index, witness)
+                else ->
+                    // we don't know how to sign this input
+                    wipTx
+            }
+        }
+    }
+}
+
+private sealed interface WalletCommand {
+    companion object {
+        object ElectrumConnected : WalletCommand
+        data class ElectrumNotification(val msg: ElectrumResponse) : WalletCommand
+        data class AddAddress(val bitcoinAddress: String, val privateKey: PrivateKey?) : WalletCommand
+    }
+}
+
+
+/**
+ * A very simple wallet that only watches one address and publishes its utxos.
+ */
+class ElectrumMiniWallet(
+    val chainHash: ByteVector32,
+    private val client: ElectrumClient,
+    private val scope: CoroutineScope
+) : CoroutineScope by scope {
+
+    // state flow with the current balance
+    private val _walletStateFlow = MutableStateFlow(WalletState(emptyMap(), emptyMap()))
+    val walletStateFlow get() = _walletStateFlow.asStateFlow()
+
+    // all currently watched script hashes and their corresponding bitcoin address
+    private var scriptHashes: Map<ByteVector32, String> = emptyMap()
+
+    // the mailbox of this "actor"
+    private val mailbox: Channel<WalletCommand> = Channel(Channel.BUFFERED)
+
+    fun addAddress(bitcoinAddress: String, privateKey: PrivateKey? = null) {
```

Comment:
If we want to make sure that this is correctly used, I think we should split it into two distinct functions:

```
fun addPay2wpkhAddress(privateKey: PrivateKey)
fun addWatchOnlyAddress(bitcoinAddress: String)
```

In the first one we would compute the address corresponding to the p2wpkh of the given private key.
This way we're sure the privateKey is always consistent with the address.

Reply:
Done in [6fcb944](https://github.com/ACINQ/lightning-kmp/pull/350/commits/6fcb944de8b901d3efbb7193f2dd8ec7c1cfe9a5). I kept the unsafe method and the API is generally sketchy, but I'd rather revisit when we know more how we are using it.